cdragontoolbox/storage.py

```diff
@@ -163,187 +119,46 @@ def stream(self, urlpath) -> RequestStreamReader:
-        if paths is None:
-            extracted_files = all_files.values()
-        else:
-            extracted_files = [all_files[path] for path in paths]
-
-        # filter already extracted files
-        extracted_files = [pf for pf in extracted_files if not os.path.isfile(self.project.storage.fspath(pf.extract_path))]
-
-        # group files by package
-        files_by_package = {}
-        for pf in extracted_files:
-            files_by_package.setdefault(pf.package, []).append(pf)
-
-        package_files_path = f"{self.path}/packages/files"
-
-        for package, files in files_by_package.items():
-            with self.project.storage.stream(f"{package_files_path}/{package}") as reader:
-                # sort files by offset to extract while streaming the bin file
-                for pkgfile in sorted(files, key=lambda f: f.offset):
-                    logger.debug("extracting %s", pkgfile.path)
-                    reader.skip_to(pkgfile.offset)
-                    fspath = self.project.storage.fspath(pkgfile.extract_path)
-                    try:
-                        os.makedirs(os.path.dirname(fspath), exist_ok=True)
-                        with open(fspath, mode='wb') as fout:
-                            if pkgfile.compressed:
-                                zobj = zlib.decompressobj(zlib.MAX_WBITS|32)
-                                writer = lambda data: fout.write(zobj.decompress(data))
-                                reader.copy(writer, pkgfile.size)
-                                fout.write(zobj.flush())
-                            else:
-                                reader.copy(fout.write, pkgfile.size)
-                    except:
-                        # remove partially downloaded files
-                        try:
-                            os.remove(fspath)
-                        except OSError:
-                            pass
-                        raise
-
-    def download(self, dry_run=False):
-        """Download project version files"""
-        logger.info("downloading project %s", self)
-        self.project.storage.download(f"{self.path}/releasemanifest", None)
-        if dry_run:
-            paths = [p for p in self.filepaths() if not os.path.isfile(self.project.storage.fspath(p))]
-            if paths:
-                logger.info("files to extract: %d", len(paths))
-            else:
-                logger.info("all files already extracted")
-        else:
-            self.extract()
+    def list_solutions(self) -> List['Solution']:
+        """List solutions present in storage"""
+        ret = []
+        base = self.fspath('solutions')
+        for name in os.listdir(base):
+            if os.path.isdir(f'{base}/{name}/releases'):
+                ret.append(Solution(self, name))
+        return ret
 
 
 class Solution:
     """
-    RADS solution
+    The Riot Application Distribution System (RADS) has two Solutions: `league_client_sln` and `lol_game_client_sln`.
```

Comment:
Technically, more solutions would be supported and the code doesn't depend on it.
I would prefer to document the concrete use case in the README (I already had some words about solutions and projects), or make it clear that the code doesn't depend on it.
Actually... there are some pieces of code that depend on the actual solution names for parsing patch versions and exporting files, but not in Solution/Project.

(Same comment for Project.)

Reply:
It looks like `SolutionVersion.patch_version` won't work for other solutions, but everything else should. I added a line in the doc string mentioning that: "These classes will likely work with other solutions, although some functionality may need to be extended."

I understand why it's useful to write these classes for general use cases, but their methods are pretty specific in some cases. I think it's nice to make it clear to the user what solutions we know we handle correctly, and therefore what their options are (unless they are doing something weird that we don't intend the classes to be used for).